### PR TITLE
Refactor DependencyModel, add unit tests, add EnvironmentAbstraction library

### PR DESCRIPTION
**Pull request by pakrym**

### CODE REVIEW

**Comment by davidfowl** [comment]
This makes me happy

**Comment by piotrpMSFT** [comment]
@pakrym can we squash and merge?

**Comment by pakrym** [comment]
No, still in progress

**Comment by pakrym** [comment]
This brings a new library `Microsoft.Extensions.EnvironmentAbstractions` which has `IFile` `IDirectory` `IEnvironment` and their default wrapper.

File: scripts/configuration/testProjects.csv [hunk 1]
```diff
@@ -5,3 +5,4 @@ dotnet-compile.UnitTests
 dotnet-build.Tests
 Microsoft.DotNet.Cli.Utils.Tests
 Microsoft.DotNet.Compiler.Common.Tests
+
```
Comment (livarcocc) [inline scripts/configuration/testProjects.csv]: What?

**Comment by pakrym** [comment]
Fixed builds

**Comment by piotrpMSFT** [comment]
@pakrym rebase and we'll merge.

**Comment by anurse** [comment]
@dotnet-bot test this please

**Comment by anurse** [comment]
Looks like MyGet timeouts

**Comment by pakrym** [comment]
@dotnet-bot test Windows_NT Release Build please
@dotnet-bot test Windows_NT Debug Build please

**Comment by pakrym** [comment]
@dotnet-bot test Windows_NT Release Build please

File: src/Microsoft.Extensions.DependencyModel/Properties/Properties.cs [hunk 2]
```diff
@@ -0,0 +1,3 @@
+﻿using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Microsoft.Extensions.DependencyModel.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
```
Comment (livarcocc) [inline src/Microsoft.Extensions.DependencyModel/Properties/Properties.cs]: This sucks.

Comment (pakrym) [inline src/Microsoft.Extensions.DependencyModel/Properties/Properties.cs]: I can make things public.

Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyModel/Properties/Properties.cs]: internal is the :imp:

File: test/Microsoft.Extensions.DependencyModel.Tests/ReferenceAssemblyResolverTests.cs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+using System;
+using System.Collections.Generic;
+using System.IO;
+using FluentAssertions;
+using Microsoft.Extensions.DependencyModel.Resolution;
+using Microsoft.Extensions.PlatformAbstractions;
+using Moq;
+using Xunit;
+
+using F = Microsoft.Extensions.DependencyModel.Tests.TestLibraryFactory;
```
Comment (davidfowl) [inline test/Microsoft.Extensions.DependencyModel.Tests/ReferenceAssemblyResolverTests.cs]: F? really?

Comment (pakrym) [inline test/Microsoft.Extensions.DependencyModel.Tests/ReferenceAssemblyResolverTests.cs]: Makes some lines 3 time shorter

File: test/Microsoft.Extensions.DependencyModel.Tests/PackageResolverTest.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+namespace Microsoft.Extensions.DependencyModel.Tests
+{
+    public class PackageResolverTest
+    {
+        private static string PackagesPath = Path.Combine("package", "directory", "location");
+
+        [Fact]
+        public void SholdUseEnvironmentVariableToGetDefaultLocation()
+        {
+
```
Comment (davidfowl) [inline test/Microsoft.Extensions.DependencyModel.Tests/PackageResolverTest.cs]: nit: why these extra spaces?

File: src/Microsoft.Extensions.EnvironmentAbstractions/project.json [hunk 5]
```diff
@@ -0,0 +1,25 @@
+{
+  "description": "Abstractions for making code that uses file system and environment testable.",
```
Comment (davidfowl) [inline src/Microsoft.Extensions.EnvironmentAbstractions/project.json]: This belongs in test not in src. It's not something we plan to ship.

Comment (pakrym) [inline src/Microsoft.Extensions.EnvironmentAbstractions/project.json]: How would test work without it?

Comment (davidfowl) [inline src/Microsoft.Extensions.EnvironmentAbstractions/project.json]: Put it somewhere that makes it seem like it's not real. Give it a name that makes it seem like it's internal. We're not putting out generic environment abstractions.

Comment (pakrym) [inline src/Microsoft.Extensions.EnvironmentAbstractions/project.json]: Ok, Micorosft.Dotnet.Internal.TestingAbstractions and make everything internal and visible only to tests?

Comment (davidfowl) [inline src/Microsoft.Extensions.EnvironmentAbstractions/project.json]: Yes, something like that.

Comment (pakrym) [inline src/Microsoft.Extensions.EnvironmentAbstractions/project.json]: Renamed the project make all of it internal, unfortunately this required putting InternalsVisibleTo in couple more place.

**Comment by pakrym** [comment]
@dotnet-bot test OSX Debug Build  please

**Comment by pakrym** [comment]
@dotnet-bot test OSX Debug Build please

File: src/Microsoft.DotNet.InternalAbstractions/Properties/Properties.cs [hunk 6]
```diff
@@ -0,0 +1,5 @@
+﻿using System.Runtime.CompilerServices;
+
```
Comment (davidfowl) [inline src/Microsoft.DotNet.InternalAbstractions/Properties/Properties.cs]: bahahhahahahaha

**Comment by davidfowl** [comment]
:shipit:

**Pull request merged**